In [87]:
import scipy
import numpy as np
import pandas as pd
import seaborn as sns
import seaborn.objects as so

from sklearn import linear_model    # Herramientas de modelos lineales
from sklearn.metrics import mean_squared_error, r2_score    # Medidas de desempeño
from sklearn.preprocessing import PolynomialFeatures    # Herramientas de polinomios

from sklearn.model_selection import train_test_split, KFold, cross_val_score

from formulaic import Formula

# Caso de estudio: calorías de alimentos

In [88]:
df_nutricion = pd.read_csv('../datasets/nutrition.csv')
df_nutricion.head()

FDC_ID                                               Item        Category  \
0  167512  Pillsbury Golden Layer Buttermilk Biscuits, Ar...  Baked Products   
1  167513  Pillsbury, Cinnamon Rolls with Icing, refriger...  Baked Products   
2  167514  Kraft Foods, Shake N Bake Original Recipe, Coa...  Baked Products   
3  167515     George Weston Bakeries, Thomas English Muffins  Baked Products   
4  167516         Waffles, buttermilk, frozen, ready-to-heat  Baked Products   

   Calorias_kcal  Proteinas_g  Carbohidratos_g  GrasaTotal_g  Colesterol_mg  \
0          307.0         5.88            41.18         13.24            0.0   
1          330.0         4.34            53.42         11.27            0.0   
2          377.0         6.10            79.80          3.70            NaN   
3          232.0         8.00            46.00          1.80            NaN   
4          273.0         6.58            41.05          9.22           15.0   

   Fibra_g  Agua_g  Alcohol_g  VitaminaC_mg  
0      1.2   35.50        NaN           NaN  
1      1.4   27.86        NaN           0.1  
2      NaN    3.20        NaN           NaN  
3      NaN   42.60        NaN           NaN  
4      2.2   40.34        0.0           0.0

En este ejemplo consideramos que los datos faltantes representan que el alimento no contiene ese ingrediente y lo convertimos a 0.

In [89]:
# Utilizamos fillna para convertir NaN a 0.
df_nutricion = df_nutricion.fillna(0)

In [90]:
df_nutricion.head()

FDC_ID                                               Item        Category  \
0  167512  Pillsbury Golden Layer Buttermilk Biscuits, Ar...  Baked Products   
1  167513  Pillsbury, Cinnamon Rolls with Icing, refriger...  Baked Products   
2  167514  Kraft Foods, Shake N Bake Original Recipe, Coa...  Baked Products   
3  167515     George Weston Bakeries, Thomas English Muffins  Baked Products   
4  167516         Waffles, buttermilk, frozen, ready-to-heat  Baked Products   

   Calorias_kcal  Proteinas_g  Carbohidratos_g  GrasaTotal_g  Colesterol_mg  \
0          307.0         5.88            41.18         13.24            0.0   
1          330.0         4.34            53.42         11.27            0.0   
2          377.0         6.10            79.80          3.70            0.0   
3          232.0         8.00            46.00          1.80            0.0   
4          273.0         6.58            41.05          9.22           15.0   

   Fibra_g  Agua_g  Alcohol_g  VitaminaC_mg  
0      1.2   35.50        0.0           0.0  
1      1.4   27.86        0.0           0.1  
2      0.0    3.20        0.0           0.0  
3      0.0   42.60        0.0           0.0  
4      2.2   40.34        0.0           0.0

Construimos las matrices X e y utilizando Formulaic

In [91]:
# No usamos intercept en este modelo
y, X = (
    Formula('Calorias_kcal ~ Proteinas_g + Carbohidratos_g + GrasaTotal_g + Colesterol_mg + Fibra_g + Agua_g + Alcohol_g + VitaminaC_mg - 1')
    .get_model_matrix(df_nutricion)
)

In [92]:
X.head() 

Proteinas_g  Carbohidratos_g  GrasaTotal_g  Colesterol_mg  Fibra_g  Agua_g  \
0         5.88            41.18         13.24            0.0      1.2   35.50   
1         4.34            53.42         11.27            0.0      1.4   27.86   
2         6.10            79.80          3.70            0.0      0.0    3.20   
3         8.00            46.00          1.80            0.0      0.0   42.60   
4         6.58            41.05          9.22           15.0      2.2   40.34   

   Alcohol_g  VitaminaC_mg  
0        0.0           0.0  
1        0.0           0.1  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

In [93]:
y.head()

Calorias_kcal
0          307.0
1          330.0
2          377.0
3          232.0
4          273.0

Antes de separar en entrenamiento y testeo, veamos los errores del modelo lineal con todos los datos 

In [94]:
modelo = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal. 
modelo.fit(X, y)   # Realizamos el ajuste

LinearRegression(fit_intercept=False)

In [95]:
modelo.coef_

array([[ 4.13762244e+00,  3.97641700e+00,  8.84901423e+00,
         5.63736883e-03, -1.68542090e+00,  1.28377434e-02,
         6.85776696e+00, -2.55226271e-02]])

Analizando los coeficientes vemos que las variables Proteinas_g, Carbohidratos_g y GrasaTotal_g son las que tienen mayor peso en el modelo.

Analizamos la "bondad" del ajuste.

In [96]:
y_pred = modelo.predict(X)
# Calculando el R^2
r2 = r2_score(y, y_pred)
print('R^2: ', r2)

# Calculando el ECM
ecm = mean_squared_error(y, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.9957808144789252
Raiz cuadarada del ECM:  10.960628490483586


A priori es un buen modelo, tenemos 7792 observaciones y obtenemos R^2 casi igual a 1 con solo 9 variables.

# Conjuntos de entrenamiento y testeo
Ajustamos el modelo separando en 80-20

In [97]:
from sklearn.model_selection import train_test_split

In [98]:
# Construimos primero las matrices X e y utilizando Formulaic (o cualquier otro método)
y, X = (
    Formula('Calorias_kcal ~ Proteinas_g + Carbohidratos_g + GrasaTotal_g + Colesterol_mg + Fibra_g + Agua_g + Alcohol_g + VitaminaC_mg - 1')
    .get_model_matrix(df_nutricion)
)

# Separamos en entrenamiento (train) y testeo (test).
# El parámetro test_size=0.2 indica que tomamos un 20% de los datos para testeo.
# El parámetro random_state=42 es lo que se denomina semilla aleatoria. Es la seed basicamente
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [99]:
X_train

Proteinas_g  Carbohidratos_g  GrasaTotal_g  Colesterol_mg  Fibra_g  \
2338         0.00            11.27          0.00            0.0      0.0   
6078        18.86             0.00         15.60           68.0      0.0   
3541        17.88             1.80          4.47          262.0      0.0   
904         19.60            57.90         13.70            0.0     37.0   
1446         3.28            23.50          0.78            0.0      2.8   
...           ...              ...           ...            ...      ...   
5226         0.22            15.32          0.07            0.0      2.0   
5390         4.00            27.54          0.18            0.0      0.0   
860          0.10            91.80          0.50            0.0      0.1   
7603        29.20             0.00         14.50           97.0      0.0   
7270         0.27            38.22          0.02            0.0      0.1   

      Agua_g  Alcohol_g  VitaminaC_mg  
2338   88.45        0.0          25.0  
6078   65.24        0.0           0.0  
3541   74.87        0.0          16.2  
904     3.00        0.0           0.0  
1446   71.79        0.0           7.2  
...      ...        ...           ...  
5226   84.13        0.0           1.0  
5390   67.91        0.0           0.0  
860     5.20        0.0           0.0  
7603   54.30        0.0           0.0  
7270   60.27        0.0           8.7  

[6234 rows x 8 columns]

In [100]:
X_test

Proteinas_g  Carbohidratos_g  GrasaTotal_g  Colesterol_mg  Fibra_g  \
1421         0.59             5.36          0.14            0.0      1.8   
5550         5.25            81.43          1.24            0.0      0.0   
5398         4.40            21.30          1.92            0.0      2.8   
1025        10.00            68.26         17.93            0.0      2.5   
101          7.99            55.10          6.40            0.0      3.6   
...           ...              ...           ...            ...      ...   
4981         8.87             4.23          2.65           62.0      0.0   
2407         0.25             0.70          0.08            0.0      0.0   
4262        11.22            80.49          1.81            0.0     13.0   
7367        27.68             0.00          7.01          100.0      0.0   
7593        29.59             0.00          8.86          109.0      0.0   

      Agua_g  Alcohol_g  VitaminaC_mg  
1421   92.99        0.0           2.0  
5550    9.96        0.0          61.3  
5398   71.61        0.0           0.0  
1025    2.30        0.0           0.0  
101    28.30        0.0           0.0  
...      ...        ...           ...  
4981   82.98        0.0           0.0  
2407   98.35        0.0           0.0  
4262    3.50        0.0           0.1  
7367   63.92        0.0           0.0  
7593   59.95        0.0           0.0  

[1559 rows x 8 columns]

### Números pseudo-aleatorios y semillas aleatorias.

Las computadoras no pueden generar números al azar, tienen algoritmos que generan números que parecen al azar denominados pseudo aleatorios. 

Los números se generan a partir de una semilla. Si corremos el codigo utilizando la misma semilla,  vamos a obtener siempre el mismo resultado.

Esto permite que el experimento sea reproducible.

### Entrenamiento

In [101]:
# Entrenamos el modelo utilizando los conjuntos de entrenamiento

modelo = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal
modelo.fit(X_train, y_train)   # Realizamos el ajuste

LinearRegression(fit_intercept=False)

### Testeo

In [102]:
# Medimos la bondad del ajuste en el conjunto de testeo

y_pred = modelo.predict(X_test)

# Calculando el R^2
r2 = r2_score(y_test, y_pred)
print('R^2: ', r2)

# Calculando el ECM
ecm = mean_squared_error(y_test, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.9964753848027386
Raiz cuadarada del ECM:  10.282313472084915


Vemos que el modelo ajusto bien en los datos de testeo, podemos confiar en el modelo obtenido.

**Ejercicio:** Utilizar distintas semillas aleatorias. ¿Se obtienen los mismos valores? ¿Se mantienen las conclusiones?

## Selección de modelos

Comparamos con un modelo utilizando solo las tres primeras variables (las de mayor peso en el modelo) y sin intercept.

In [103]:
y, X = (
    Formula('Calorias_kcal ~ Proteinas_g + Carbohidratos_g + GrasaTotal_g - 1')
    .get_model_matrix(df_nutricion)
)
X.head()

Proteinas_g  Carbohidratos_g  GrasaTotal_g
0         5.88            41.18         13.24
1         4.34            53.42         11.27
2         6.10            79.80          3.70
3         8.00            46.00          1.80
4         6.58            41.05          9.22

In [104]:
# Separamos en entrenamiento (train) y testeo (test).
# El parámetro test_size=0.2 indica que tomamos un 20% de los datos para testeo.
# El parámetro random_state=42 es lo que se denomina semilla aleatoria. 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [105]:
modelo = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
modelo.fit(X_train, y_train)   # Realiza
print("Coeficientes:", modelo.coef_)

y_pred = modelo.predict(X_test)
# Calculando el R^2
r2 = r2_score(y_test, y_pred)
print('R^2: ', r2)

# Calculando el ECM
ecm = mean_squared_error(y_test, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

Coeficientes: [[4.1353808  3.85362735 8.84446623]]
R^2:  0.9906661296314964
Raiz cuadarada del ECM:  16.7326939492239


Vemos que el modelo es un poco peor pero mucho más simple. 

**Conclusión rápida:** El modelo con 3 variables es útil para una cuenta rápida, pero si necesitamos una mayor precisión podemos utilizar el modelo completo.

**Ejercicio.** Buscar en recursos en-líne]a la fórmula usualmente utilizada para el cálculo de calorías y comparar con la fórmula que obtuvimos.

# Caso de estudio: rendimiento del suelo

In [106]:
df_rendimiento = pd.read_csv('../datasets/rendimiento.csv')
df_rendimiento.head()

Ano             Localidad   x1   x2   x3  x4   x5   x6   x7 x8  ...  \
0  2007  San Antonio de Areco  6.2  3.1  6.0  26  4.3  MAP  100  S  ...   
1  2007  San Antonio de Areco  6.2  3.1  6.0  26  4.3  MAP  100  S  ...   
2  2007  San Antonio de Areco  6.2  3.1  6.0  26  4.3  MAP  100  S  ...   
3  2007  San Antonio de Areco  6.2  3.1  6.0  26  4.3  MAP  100  S  ...   
4  2007  San Antonio de Areco  6.2  3.1  6.0  26  4.3  MAP  140  S  ...   

           x12    x13  x14    x15  x16   x17  x18     x19  x20     y  
0  Baguette 11  5-Jun  260  Z 3.7  4.0  35.0  4.0  24-Oct  621  5666  
1  Baguette 11  5-Jun  260  Z 3.7  4.0  35.0  4.0  24-Oct  673  5874  
2  Baguette 11  5-Jun  249  Z 3.7  4.0  35.0  4.0  24-Oct  671  5862  
3  Baguette 11  5-Jun  252  Z 3.7  4.0  35.0  4.0  24-Oct  721  6055  
4  Baguette 11  5-Jun  272  Z 3.7  4.0  35.0  4.0  24-Oct  656  5660  

[5 rows x 23 columns]

**Ejercicio.**

Proponer un modelo para predecir el rendimiento de una hectárea cultivada en función de algunas características del lugar y los fertilizantes utilizados.

In [107]:
df_rendimiento.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 23 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Ano        156 non-null    int64  
 1   Localidad  156 non-null    object 
 2   x1         156 non-null    float64
 3   x2         156 non-null    float64
 4   x3         156 non-null    float64
 5   x4         156 non-null    int64  
 6   x5         156 non-null    float64
 7   x6         156 non-null    object 
 8   x7         156 non-null    int64  
 9   x8         156 non-null    object 
 10  x9         156 non-null    int64  
 11  x10        156 non-null    int64  
 12  x11        156 non-null    float64
 13  x12        156 non-null    object 
 14  x13        156 non-null    object 
 15  x14        156 non-null    int64  
 16  x15        156 non-null    object 
 17  x16        156 non-null    float64
 18  x17        156 non-null    float64
 19  x18        156 non-null    float64
 20  x19       

In [108]:
# Acá están todas las variables
"Rendimiento ~ pH + MO + P + N + S + P_Ferti + P_kgha + S_Fuente + S_kgha + Ndisp + Palc + Variedad + FdS + plantas + momento + IncSep + IncMam + IncRoya + FdD + Espigas"

'Rendimiento ~ pH + MO + P + N + S + P_Ferti + P_kgha + S_Fuente + S_kgha + Ndisp + Palc + Variedad + FdS + plantas + momento + IncSep + IncMam + IncRoya + FdD + Espigas'

# Cuatro niveles de selección de modelos
Queremos estimar los costos de salud que tendrá un cliente de una prepaga en función de algunas variables de la persona.

In [127]:
df_salud = pd.read_csv('../datasets/insurance.csv')
df_salud.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

La variable respuesta es `charges` y el resto son variables explicativas.

**Pregunta:** ¿Cuáles son variables numércias? ¿Cuáles son categóricas? Las variables categóricas son: ¿binarias, nominales, ordinales?

In [128]:
df_salud.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


Las variables numericas son age, bmi, children, charges.

Las variables sex, smoker son binarias, y la variable region es nominal.

**Pregunta:** ¿Cuántas categorías tiene la variable `region`?

In [129]:
# Otro comando útil para analizar las variables numéricas
df_salud.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

Por ejemplo, en la variable `charges` vemos que el promedio es 13270, y el máximo es 63770. Esto podría indica la presencia de outliers, que dificultan el modelo. ¿Cómo podemos visualizar los outliers?

In [130]:
df_salud.iloc[df_salud["charges"].argmax()]

age                  54
sex              female
bmi               47.41
children              0
smoker              yes
region        southeast
charges     63770.42801
Name: 543, dtype: object

the usual suspects xD

En este ejemplo dejamos los outliers. Ejercicio: ¿cambian las conclusiones si eliminamos los outliers?

## Nivel 1: entrenamos y testeamos el modelo usando todos los datos

In [131]:
y, X = (
    Formula('charges ~ age + sex + bmi + children + smoker + region-1')
    .get_model_matrix(df_salud)
)
X.head()

age  sex[female]  sex[male]     bmi  children  smoker[T.yes]  \
0   19            1          0  27.900         0              1   
1   18            0          1  33.770         1              0   
2   28            0          1  33.000         3              0   
3   33            0          1  22.705         0              0   
4   32            0          1  28.880         0              0   

   region[T.northwest]  region[T.southeast]  region[T.southwest]  
0                    0                    0                    1  
1                    0                    1                    0  
2                    0                    1                    0  
3                    1                    0                    0  
4                    1                    0                    0

In [132]:
# Podemos ver la correlación entre las distintas variables (corresponde al R de un modelo lineal y=ax+b)
pd.concat([X,y], axis = 1).corr()

age  sex[female]  sex[male]       bmi  children  \
age                  1.000000     0.020856  -0.020856  0.109272  0.042469   
sex[female]          0.020856     1.000000  -1.000000 -0.046371 -0.017163   
sex[male]           -0.020856    -1.000000   1.000000  0.046371  0.017163   
bmi                  0.109272    -0.046371   0.046371  1.000000  0.012759   
children             0.042469    -0.017163   0.017163  0.012759  1.000000   
smoker[T.yes]       -0.025019    -0.076185   0.076185  0.003750  0.007673   
region[T.northwest] -0.000407     0.011156  -0.011156 -0.135996  0.024806   
region[T.southeast] -0.011642    -0.017117   0.017117  0.270025 -0.023066   
region[T.southwest]  0.010016     0.004184  -0.004184 -0.006205  0.021914   
charges              0.299008    -0.057292   0.057292  0.198341  0.067998   

                     smoker[T.yes]  region[T.northwest]  region[T.southeast]  \
age                      -0.025019            -0.000407            -0.011642   
sex[female]              -0.076185             0.011156            -0.017117   
sex[male]                 0.076185            -0.011156             0.017117   
bmi                       0.003750            -0.135996             0.270025   
children                  0.007673             0.024806            -0.023066   
smoker[T.yes]             1.000000            -0.036945             0.068498   
region[T.northwest]      -0.036945             1.000000            -0.346265   
region[T.southeast]       0.068498            -0.346265             1.000000   
region[T.southwest]      -0.036945            -0.320829            -0.346265   
charges                   0.787251            -0.039905             0.073982   

                     region[T.southwest]   charges  
age                             0.010016  0.299008  
sex[female]                     0.004184 -0.057292  
sex[male]                      -0.004184  0.057292  
bmi                            -0.006205  0.198341  
children                        0.021914  0.067998  
smoker[T.yes]                  -0.036945  0.787251  
region[T.northwest]            -0.320829 -0.039905  
region[T.southeast]            -0.346265  0.073982  
region[T.southwest]             1.000000 -0.043210  
charges                        -0.043210  1.000000

In [133]:
# Ajustamos el modelo
modelo = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
modelo.fit(X, y)   # Realiza
print("Coeficientes:", modelo.coef_)

Coeficientes: [[   256.85635254 -11938.53857617 -12069.85293556    339.19345361
     475.50054515  23848.53454191   -352.96389942  -1035.02204939
    -960.0509913 ]]


In [134]:
# Predicciones
y_pred = modelo.predict(X)

# Bondad del ajuste
r2 = r2_score(y, y_pred)
print('R^2: ', r2)
ecm = mean_squared_error(y, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.7509130345985205
Raiz cuadarada del ECM:  6041.679651174452


## Nivel 2: separamos en entrenamiento y testeo

In [135]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [138]:
y_train

charges
560    9193.83850
1285   8534.67180
1142  27117.99378
969    8596.82780
486   12475.35130
...           ...
1095   4561.18850
1130   8582.30230
1294  11931.12525
860   46113.51100
1126  10214.63600

[1070 rows x 1 columns]

In [137]:
# Ajustamos el modelo
modelo = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
modelo.fit(X_train, y_train)   # Realiza

# Predicciones
y_pred = modelo.predict(X_test)

# Bondad del ajuste
r2 = r2_score(y_test, y_pred)
print('R^2: ', r2)
ecm = mean_squared_error(y_test, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.7835929767120724
Raiz cuadarada del ECM:  5796.284659276272


**Pregunta:** Con la semilla aleatoria 42 bajó el ECM. Si cambiamos la semilla por ejemplo a 4?

## Nivel 3: separamos en entrenamiento, validación y testeo
#### Paso 1: separamos en entrenamiento y testeo el dataframe original

In [120]:
df_train, df_test = train_test_split(df_salud, test_size=0.2, random_state=42)
df_train.shape

(1070, 7)

Podemos también primero transformar las variables y después separar en entrenamiento y testeo, pero tenemos que tener cuidado y evitar usar en el entrenamiento datos del conjunto de testeo.

Por ejemplo, si queremos normalizar una variable mediante StandardScaler, tenemos que calcular los parámetros de la normalización en los datos de entrenamiento y aplicar esa misma transformación en los datos de testeo.

#### Paso 2A: definimos un primer modelo y separamos el dataset df_train en entrenamiento y validación para entrenar el modelo.

In [121]:
formula1 = 'charges ~ age + sex + bmi + children + smoker + region'
y1, X1 = (
    Formula(formula1)
    .get_model_matrix(df_train)
)

In [122]:
X_train_1, X_val_1, y_train_1, y_val_1 = train_test_split(X1, y1, test_size=0.2, random_state=42)
X_train_1.shape

(856, 9)

In [123]:
modelo1 = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
modelo1.fit(X_train_1, y_train_1)   # Realizamos el ajuste

# Predicciones
y_pred = modelo1.predict(X_val_1)

# Bondad del ajuste
r2 = r2_score(y_val_1, y_pred)
print('R^2: ', r2)
ecm = mean_squared_error(y_val_1, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.7268029806501225
Raiz cuadarada del ECM:  6602.039502172559


#### Paso 2B: definimos otro modelo y repetimos el paso 2A

In [124]:
formula2 = 'charges ~ age + bmi + children + smoker'
y2, X2 = (
    Formula(formula2)
    .get_model_matrix(df_train)
)
X_train_2, X_val_2, y_train_2, y_val_2 = train_test_split(X2, y2, test_size=0.2, random_state=42)
X_train_2.shape

(856, 5)

In [125]:
modelo1 = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
modelo1.fit(X_train_2, y_train_2)   # Realizamos el ajuste

# Predicciones
y_pred_2 = modelo1.predict(X_val_2)

# Bondad del ajuste
r2 = r2_score(y_val_2, y_pred_2)
print('R^2: ', r2)
ecm = mean_squared_error(y_val_2, y_pred_2)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.7263709097118607
Raiz cuadarada del ECM:  6607.25812140838


De los modelos probados, nos quedamos con el de menor RECM.

#### Paso 3: analizamos como funciona el modelo en el conjunto de testeo.

**Importante:** Para esto, entrenamos el modelo ganador utilizando **TODOS** los datos de entrenamiento (el modelo es la fórmula, no los coeficientes).

**Recordar:** mientras mas datos usamos para entrenar, mejor!

In [126]:
# Ajustamos nuestro modelo ganador en TODO el conjunto de entrenamiento. 
modelo1.fit(X1, y1)

# Realizamos las mismas transformaciones en el conjunto de testeo
y_test, X_test = (
    Formula(formula1)
    .get_model_matrix(???)
)

# Predicciones
y_pred = modelo1.predict(???)

# Bondad del ajuste
r2 = r2_score(y_test, y_pred)
print('R^2: ', r2)
ecm = mean_squared_error(y_test, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

SyntaxError: invalid syntax (497964493.py, line 7)

## Nivel 4: separamos en entrenamiento y testeo, y hacemos validación cruzada en el conjunto de entrenamiento.
### Paso 1: separamos en entrenamiento y testeo el dataframe original

In [ ]:
# La misma separación del Nivel 3
df_train, df_test = train_test_split(df_salud, test_size=0.2, random_state=42)
df_train.shape

(1070, 7)

#### Paso 2A: definimos un primer modelo y lo ajustamos por validación cruzada en el conjunto de entrenamiento.

In [ ]:
formula1 = 'charges ~ age + sex + bmi + children + smoker + region'
y1, X1 = (
    Formula(formula1)
    .get_model_matrix(df_train)
)
X1.head()

Intercept  age  sex[T.male]    bmi  children  smoker[T.yes]  \
560         1.0   46            0  19.95         2              0   
1285        1.0   47            0  24.32         0              0   
1142        1.0   52            0  24.86         0              0   
969         1.0   39            0  34.32         5              0   
486         1.0   54            0  21.47         3              0   

      region[T.northwest]  region[T.southeast]  region[T.southwest]  
560                     1                    0                    0  
1285                    0                    0                    0  
1142                    0                    1                    0  
969                     0                    1                    0  
486                     1                    0                    0

In [ ]:
# Definimos los subconjuntos para la validación cruzada.
# Utilizamos KFold de sklearn
cv = KFold(n_splits=5, random_state=42, shuffle=True)

## Generadores e iteradores perezosos en Python
Nos detenemos un momento para entender qué nos devuelve KFold

In [ ]:
# Esto solo nos muestra las opciones que utilizamos
cv

KFold(n_splits=5, random_state=42, shuffle=True)

In [ ]:
# La forma de utilizado es a través del método split
pliegos = cv.split(X1)
pliegos

<generator object _BaseKFold.split at 0x7fecedcab560>

`split` nos devuelve un "generador", esto es un **iterador perezoso** (lazy iterator).

Los iteradores perezosos son objetos que se pueden recorrer como una lista. 

Sin embargo, a diferencia de las listas, los iteradores perezosos no almacenan su contenido en la memoria, lo van generando a medida que lo necesitamos.


In [ ]:
# Podemos acceder a los elementos a través de la función next
next(pliegos)

(array([   0,    1,    4,    5,    6,    7,    8,    9,   11,   13,   14,
          15,   16,   17,   18,   19,   20,   21,   22,   24,   25,   26,
          27,   28,   29,   32,   33,   34,   35,   36,   37,   38,   40,
          41,   42,   43,   44,   45,   46,   47,   48,   50,   52,   53,
          55,   56,   57,   60,   61,   62,   64,   65,   68,   69,   71,
          72,   73,   74,   75,   77,   79,   80,   81,   82,   84,   85,
          87,   89,   90,   91,   92,   93,   94,   95,   97,   98,   99,
         102,  103,  104,  105,  106,  108,  110,  111,  112,  114,  115,
         116,  117,  118,  119,  120,  121,  122,  123,  124,  125,  126,
         127,  128,  129,  130,  131,  132,  133,  134,  135,  137,  138,
         140,  142,  143,  144,  145,  146,  147,  148,  149,  150,  151,
         152,  153,  154,  155,  156,  157,  159,  160,  161,  162,  163,
         164,  165,  166,  167,  169,  170,  171,  172,  173,  175,  176,
         177,  179,  180,  181,  182, 

In [ ]:
# Pero lo mas común es utilizarlos en un ciclo:
pliegos = cv.split(X1)
for train_index, test_index in pliegos:
    print(test_index[0:10])

[ 2  3 10 12 23 30 31 39 49 51]
[ 0  5  7  9 25 29 33 43 44 55]
[ 6 11 15 18 22 24 28 41 42 46]
[ 8 16 17 19 26 36 37 38 45 48]
[ 1  4 13 14 20 21 27 32 34 35]


In [ ]:
# Ahora no quedo nada, ya generó todo lo que tenía para generar
next(pliegos)

StopIteration: 

In [ ]:
# Acá tampoco hay nada...
for train_index, test_index in pliegos:
    print(test_index[0:10])

#### Volvemos al Paso 2A

In [ ]:
# Para seleccionar algunas filas dados los índices, utilizamos iloc (lo vimos en la clase 2)
for train_index, val_index in cv.split(X1):
    X_train, X_val, y_train, y_val = X1.iloc[train_index], X1.iloc[val_index], y1.iloc[train_index], y1.iloc[val_index]
    
    # Acá tenemos que hacer el ajuste y la predicción para cada pliego

Agregamos el codigo para ajuste y predicción

In [ ]:
modelo1 = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
rmse1 = np.zeros(cv.get_n_splits())  # Vamos a guardar el error en cada pliego

ind = 0

# Para seleccionar algunas filas dados los índices, utilizamos iloc (lo vimos en la clase 2)
for train_index, test_index in cv.split(X1):
    X_train, X_val, y_train, y_val = X1.iloc[train_index], X1.iloc[val_index], y1.iloc[train_index], y1.iloc[val_index]
    modelo1.fit(X_train, y_train)
    
    y_pred = modelo1.predict(X_val)
    rmse1[ind] = np.sqrt(mean_squared_error(y_val, y_pred))
    ind = ind + 1

In [ ]:
print(rmse1)

[5603.6714797  5601.06002551 5601.94864651 5593.78511835 5621.49155797]


In [ ]:
print(rmse1.mean())  # Este es el valor que queremos minimizar

5604.391365607384


#### Paso 2B: definimos otro modelo y repetimos el paso 2A

In [ ]:
formula2 = 'charges ~ age + bmi + children + region + smoker'
y2, X2 = (
    Formula(formula2)
    .get_model_matrix(df_train)
)

cv = KFold(n_splits=5, random_state=42, shuffle=True)  # No es necesario definirlo nuevamente, solo para recordar que era.

modelo2 = linear_model.LinearRegression(fit_intercept = False)    # Inicializamos un modelo de Regresion Lineal sin intercept
rmse2 = np.zeros(cv.get_n_splits())  # Vamos a guardar el error en cada pliego

ind = 0

# Para seleccionar algunas filas dados los índices, utilizamos iloc (lo vimos en la clase 2)
for train_index, test_index in cv.split(X2):
    X_train, X_val, y_train, y_val = X2.iloc[train_index], X2.iloc[val_index], y2.iloc[train_index], y2.iloc[val_index]
    modelo2.fit(X_train, y_train)
    
    y_pred = modelo2.predict(X_val)
    rmse2[ind] = np.sqrt(mean_squared_error(y_val, y_pred))
    ind = ind + 1

In [ ]:
print(rmse2)
print(rmse2.mean())  # Este es el valor que queremos minimizar

[5601.5478229  5600.54600166 5601.12884663 5593.12637971 5621.34963652]
5603.539737483336


De los modelos probados, nos quedamos con el de menor RECM. 

#### Paso 3: analizamos como funciona el modelo en el conjunto de testeo.

Copiamos el mismo código del paso 3 del nivel 3.

In [ ]:
# Ajustamos nuestro modelo ganador en TODO el conjunto de entrenamiento. 
modelo2.fit(X2, y2)

# Realizamos las mismas transformaciones en el conjunto de testeo
y_test, X_test = (
    Formula(formula2)
    .get_model_matrix(df_test)
)

# Predicciones
y_pred = modelo2.predict(X_test)

# Bondad del ajuste
r2 = r2_score(y_test, y_pred)
print('R^2: ', r2)
ecm = mean_squared_error(y_test, y_pred)
print('Raiz cuadarada del ECM: ', np.sqrt(ecm))

R^2:  0.7835569786290855
Raiz cuadarada del ECM:  5796.766728796432


## Ejercicio

Repetir los ejercicios de la Prática 6 (Modelo Lineal Multivariado), proponiendo distintos modelos lineales y utilizando validación cruzada para seleccionar el mejor modelo.
